packages/jsii/lib/assembler.ts

```diff
@@ -148,11 +146,38 @@ export class Assembler implements Emitter {
      * Defer checks for after the program has been entirely processed; useful for verifying type references that may not
      * have been discovered yet, and verifying properties about them.
      *
+     * @param fqn FQN of the current type.
+     * @param deps List of FQNs of types this callback depends on. All deferreds for all
      * @param cb the function to be called in a deferred way. It will be bound with ``this``, so it can depend on using
      *           ``this``.
      */
-    private _defer(cb: () => void) {
-        this._deferred.push(cb.bind(this));
+    private _defer(fqn: string, depFqns: string[], cb: () => void) {
```

Comment:
I don't agree that it's sketchy--it's running some analysis in a guaranteed order. I was debating whether to rewrite the whole assembler into two phases, first collect all types and relationships between types, then do the regular processing in order. 

Part of the processing doesn't require the base types though, so I ultimately decided it would be a lower-impact change to leave the standalone analysis in the first pass and only do the analysis that requires base types in the second pass--and the "deferred" mechanism could be extended easily enough to accomodate the ordering required for that second pass.

Now, whether or not it should be inlined is another case. In principle the ordered analysis method is more general than it's being used for right now. It doesn't NEED to work on `NamedTypeReferences` that are `dereferenced`... just so happens that all places where it's being used use that pattern, so I extracted that out again. I don't think it hurts to show there are two layers to it.